corehq/apps/data_interfaces/models.py

```diff
@@ -1214,6 +1324,101 @@ def to_dict(self):
         }
 
 
+class CaseDuplicateNew(models.Model):
+    id = models.BigAutoField(primary_key=True)
+    case_id = models.CharField(max_length=126, db_index=True)
+    action = models.ForeignKey("CaseDeduplicationActionDefinition", on_delete=models.CASCADE)
+    match_values = models.CharField(max_length=256)
+
+    class Meta:
+        db_table = "data_interfaces_caseduplicate_new"
+        unique_together = ('case_id', 'action')
+        indexes = [
+            models.Index(fields=['match_values', 'action_id'])
+        ]
+
+    def __str__(self):
+        return (
+            f"CaseDuplicateNew("
+            f"case_id={self.case_id}, action_id={self.action_id}, match_values={self.match_values})"
+        )
+
+    def delete(self, *args, check_for_orphans=True, **kwargs):
+        with transaction.atomic():
+            if check_for_orphans:
+                other_records = CaseDuplicateNew.objects.filter(
+                    action=self.action, match_values=self.match_values).exclude(case_id=self.case_id)[:2]
+
+                if other_records.count() == 1:
+                    # This will be orphaned when the current record is deleted, so delete it as well
+                    other_records[0].delete(check_for_orphans=False)
```

Comment:
An unnecessary `SAVEPOINT` will be created on this recursive delete due to the nested `transaction.atomic()`. What do you think of restructuring this to only use a transaction if this extra delete will be executed, and otherwise not?

Reply:
I think [this post](https://stackoverflow.com/a/5982654) backs up my assumption that code like the following:
```
SELECT * FROM SOME_TABLE;
UPDATE VALUE FROM SOME_TABLE WHERE ....;
```
operates differently depending on whether or not those lines are executed in a transaction. Without an enclosing transaction, the values in `SOME_TABLE` can be modified in between the select and updates clauses, whereas with a transaction, they can't (assuming the isolation level is sufficiently strict). I don't think any performance gains are worth the tradeoff in reliability -- I want the determination on whether or not this delete will orphan records to be current when I actually delete the data, which is why the select is also included in the transaction, not just the deletes.

Reply:
I was suggesting moving the transaction block under `if check_for_orphans:` and using `transaction.atomic(savepoint=False)` (the same technique used in Django's delete implementation). See suggestion below.